tracing-appender/src/non_blocking.rs

```diff
@@ -144,3 +144,155 @@ impl Drop for WorkerGuard {
+    }
+
+    impl MockWriter {
+        fn new(capacity: usize) -> (Self, mpsc::Receiver<String>) {
+            let (tx, rx) = mpsc::sync_channel(capacity);
+            (Self { tx }, rx)
+        }
+    }
+
+    impl std::io::Write for MockWriter {
+        fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+            let buf_len = buf.len();
+            let _ = self.tx.send(String::from_utf8_lossy(buf).to_string());
+            Ok(buf_len)
+        }
+
+        fn flush(&mut self) -> std::io::Result<()> {
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn backpressure_exerted() {
+        let (mock_writer, rx) = MockWriter::new(1);
+
+        let (mut non_blocking, _guard) = self::NonBlockingBuilder::default()
+            .lossy(false)
+            .buffered_lines_limit(1)
+            .finish(mock_writer);
+
+        let error_count = non_blocking.error_counter();
+
+        non_blocking.write_all(b"Hello").expect("Failed to write");
+        assert_eq!(0, error_count.load(Ordering::Relaxed));
+
+        let handle = thread::spawn(move || {
+            non_blocking.write_all(b", World").expect("Failed to write");
+        });
+
+        // Sleep a little to ensure previously spawned thread get's blocked on write.
+        thread::sleep(Duration::from_millis(100));
+        // We should not drop logs when blocked.
+        assert_eq!(0, error_count.load(Ordering::Relaxed));
+
+        // Read the first message to unblock sender.
+        let mut line = rx.recv().unwrap();
+        assert_eq!(line, "Hello");
+
+        // Wait for thread to finish.
+        let _ = handle.join();
+
+        // Thread has joined, we should be able to read the message it sent.
+        line = rx.recv().unwrap();
+        assert_eq!(line, ", World");
+    }
+
+    #[test]
+    fn logs_dropped_if_lossy() {
+        let (mock_writer, rx) = MockWriter::new(1);
+
+        let (mut non_blocking, _guard) = self::NonBlockingBuilder::default()
+            .lossy(true)
+            .buffered_lines_limit(1)
+            .finish(mock_writer);
+
+        let error_count = non_blocking.error_counter();
+
+        non_blocking.write_all(b"Hello").expect("Failed to write");
+        assert_eq!(0, error_count.load(Ordering::Relaxed));
+
+        non_blocking.write_all(b", World").expect("Failed to write");
+        assert_eq!(1, error_count.load(Ordering::Relaxed));
+
+        non_blocking.write_all(b".").expect("Failed to write");
+        assert_eq!(2, error_count.load(Ordering::Relaxed));
+
+        // Allow a line to be written
+        let line = rx.recv().unwrap();
+        assert_eq!(line, "Hello");
+
+        // Now, there is once again capacity in the buffer.
+        non_blocking
+            .write_all(b"Universe")
+            .expect("Failed to write");
+        assert_eq!(2, error_count.load(Ordering::Relaxed));
+
+        let line = rx.recv().unwrap();
+        assert_eq!(line, "Universe");
+        drop(non_blocking);
```

Comment:
Can you add a comment here, or on [`NonBlocking`] explaining the importance of `dropping` something? Quoting what you said in your review comment:

> Looks like for the case when non blocking is lossy, I must drop non_blocking at the end of the test otherwise it will hang some times.

Reply:
I'll add a comment in the tests and also in the documentation PR concerning this.

Reply:
This seems like a bug. I think we should _fix_ this rather than documenting it.

Reply:
Are we confident that this occurs outside of this particular test, though? If not, it could be a consequence of how we're using the bounded channel as a mock writer.

Reply:
This is problem occurs in both tests where non blocking is set to lossy. From what I've found so far is that we seem to hang at `let line = rx.recv().unwrap();`. I believe it's the way we are using the bounded channel but not sure what exactly yet. I'll spend some more time tomorrow trying to debug it.

Reply:
The tests do actually fully pass all assert statements, but this behaviour is only encountered when there's a usage of `rx`. Is it possible the receiver is somehow preventing the worker thread from getting dropped and thus the test just spins at the end? 

I'm not sure if there are some tools available to us to see what the tests are doing while hanging.

Reply:
> The tests do actually fully pass all assert statements, but this behaviour is only encountered when there's a usage of rx. Is it possible the receiver is somehow preventing the worker thread from getting dropped and thus the test just spins at the end?

Not sure. Do we want to block the review on this test, or do we want to remove this test wholesale, @hawkw?

> I'm not sure if there are some tools available to us to see what the tests are doing while hanging.

`tracing`? :P

But seriously, Xcode's instruments are _really_ solid; I'd give it a shot if you have a Mac handy. I'm not sure there's anything as easy to use on an AL2012 box.

Reply:
I don't think we should remove the tests as they are still "safe" as long as we drop `non_blocking` at the end.

Reply:
I'd like to take a minute to look at this. I think I understand what's going on, and there should be other solutions.

Reply:
The latest commit fixes the problem. Thanks for identifiying the race that was happening @hawkw